### PR TITLE
Add a `@review` command

### PR DESCRIPTION
**Pull request by dustinbyrne**
This pull request introduces a new `@review` command, enabling a seamless integration for conducting code reviews based on git diffs. This feature helps automate and streamline the code review process for pull requests by providing a structured analysis of code changes. 

#### Features

- **Command Integration**: 
  - The `@review` command hooks into the pre-existing explain functionality to identify user requests for code review, then generates the necessary git diff data.
  
- **Git Diff Handling**: 
  - The command efficiently extracts git diffs from the specified range (e.g., `base..HEAD`) and integrates this into the user context for review.
  
- **Contextual Analysis**:
  - Conducts a robust analysis of diffs in two rounds:
    1. An initial chain-of-thought review focusing on specified areas of concern.
    2. A final compiled summary that includes insights from both previous analyses and context.

- **Feedback Focus Areas**:
  - The review focuses on correctness, code quality, performance, security, documentation, testing, compatibility, and design. The feedback is rendered in a user-friendly, structured format that highlights actionable improvements.

- **Extension and Maintenance**:
  - Adds new classes and functions for handling review logic, making the codebase extensible for future improvements or new features.
  - Includes unit tests to ensure the reliability of new features and to maintain the integrity of the codebase.

#### Implementation Details

- **Modular Design**: 
  - Introduced a `ReviewCommand` class implementing the `Command` interface, encapsulating the logic for managing the review process. 
  - The `handleReview` function processes the request to extract git diffs and integrates seamlessly with existing code infrastructure.

- **Testing**:
  - Comprehensive unit tests verify that various scenarios, such as the absence of diffs and incorrect command usage, are handled gracefully.

### CODE REVIEW

**Comment by dustinbyrne** [comment]
It's worth mentioning that I went through a couple of rounds of changes using this command before submitting this pull request. With everything resolved, further `@review` requests come back clean:
> ### Summary of Code Review
> 
> #### Correctness
> - The `@review` command is correctly integrated and functions as intended. Ensure edge cases, such as unusual git states, are handled gracefully.
> 
> #### Code Quality
> - The code is well-structured and readable. Consider adding more inline comments for complex logic to aid future maintainers.
> 
> #### Performance
> - The current implementation is efficient. Monitor performance with large diffs and consider optimizations if necessary.
> 
> #### Security
> - The use of `execFile` is appropriate for security. Ensure user input is sanitized to prevent command injection vulnerabilities.
> 
> #### Documentation
> - Update user-facing documentation to include the new `@review` command, detailing its usage and options.
> 
> #### Testing
> - The tests are comprehensive. Consider adding more edge case scenarios to ensure robustness against unexpected inputs or states.
> 
> #### Compatibility
> - The changes are backward compatible. Continue verifying integration with other components to prevent regressions.
> 
> #### Design and Architecture
> - The design is sound and fits well within the existing architecture. The use of a command pattern is appropriate and facilitates future extensions.
> 
> Overall, the changes are well-implemented and align with best practices. Addressing the minor suggestions above will further enhance the quality and robustness of the code.

**Comment by dustinbyrne** [comment]
For completeness, here's one of my earliest `@review` outputs. These were all things that I addressed.

> ### Summary of Code Review
> 
> #### Correctness
> - Ensure robust handling of potential failures in the `git` command execution within the `handleReview` function. Consider implementing error handling for `execSync` to manage exceptions gracefully.
> 
> #### Code Quality
> - Improve code readability by adding comments to the `handleReview` function and `ReviewCommand` class. This will aid in understanding the logic and purpose of key sections.
> 
> #### Performance
> - Evaluate the use of `execSync` for processing git diffs. Consider adopting asynchronous methods to prevent blocking the main thread, especially when dealing with large diffs.
> 
> #### Security
> - Validate and sanitize inputs used in `execSync` to mitigate the risk of command injection vulnerabilities. This is crucial for maintaining the security of the application.
> 
> #### Documentation
> - Enhance inline documentation to provide clarity on the functionality and purpose of the code. This will improve maintainability and ease of understanding for future developers.
> 
> #### Testing
> - Expand test coverage to include scenarios where the `git` command might fail or return unexpected results. This will ensure the robustness of the application under various conditions.
> 
> Overall, the changes are well-structured and introduce valuable functionality. Addressing the highlighted areas will further enhance the quality, security, and performance of the code.

File: packages/cli/src/rpc/explain/review.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+export default async function handleReview(
+  question: string,
+  userContext?: UserContext.Context
+): Promise<{ applied: boolean; userContext?: UserContext.Context }> {
+  const [mode] = question.split(/\s+/g);
+  if (mode !== '@review') return { applied: false };
+
+  const result = parseOptions(question);
+  const base = result.options.stringValue('base', 'main');
+  const cwd = result.options.stringValue('project', configuration().projectDirectories[0]);
```
Comment (kgilpin) [inline packages/cli/src/rpc/explain/review.ts:40]: I think it's important to diff all the project directories that are configured. Admittedly, the `/base` could be different across different projects and then the results won't really seem right; but the default "base" detector can be applied to each project. 

If a project directory is not git-controlled, then it can be ignored for `@review`.

File: packages/cli/src/rpc/explain/review.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+export default async function handleReview(
+  question: string,
+  userContext?: UserContext.Context
+): Promise<{ applied: boolean; userContext?: UserContext.Context }> {
+  const [mode] = question.split(/\s+/g);
+  if (mode !== '@review') return { applied: false };
+
+  const result = parseOptions(question);
+  const base = result.options.stringValue('base', 'main');
```
Comment (kgilpin) [inline packages/cli/src/rpc/explain/review.ts:39]: Just pointing out that parsing options and applying them within the CLI is a bit different way for this to work. Can't think of any reason not to do it this way though...

File: packages/cli/src/rpc/explain/review.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  return {
+    applied: true,
+    userContext: [
+      ...(typeof userContext === 'string'
+        ? [{ content: userContext, type: 'code-selection' } as UserContext.CodeSelectionItem]
+        : userContext ?? []),
+      {
+        type: 'code-snippet',
+        location: 'git diff',
+        content: await exec('git', ['log', '-p', '--full-diff', `${base}..HEAD`], { cwd }),
```
Comment (kgilpin) [inline packages/cli/src/rpc/explain/review.ts:50]: When I manually create the diff and then pass it into Navie, it's quite possible for there to be an overflow of the token limit; this is especially likely when big file changes like yarn.lock are included. What do you think we should do about really big diffs?

File: packages/navie/src/commands/review-command.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+- **Correctness**: Identify any bugs, logical errors, or edge cases that may cause the code to fail.
+- **Code Quality**: Assess readability, maintainability, and adherence to coding standards and style guidelines.
+- **Performance**: Point out any inefficiencies or opportunities for optimization.
+- **Security**: Highlight potential security vulnerabilities or risks introduced by the changes.
+- **Documentation**: Ensure that the code is well-documented where necessary, including comments and documentation updates.
+- **Testing**: Check if the changes are adequately covered by unit tests and suggest additional tests if needed.
+- **Compatibility**: Consider the impact on existing functionality, including backward compatibility and integration with other system components.
+- **Design and Architecture**: Evaluate whether the code follows good design principles and fits well within the overall architecture.`;
```
Comment (kgilpin) [inline packages/navie/src/commands/review-command.ts:18]: Is there a way to make the output compact when Navie decides that a particular section is OK as is? Reading `r/codereview` (example: https://www.reddit.com/r/codereview/comments/1gaj3vg/what_ai_functionality_would_you_want_in_a_code/) a primary complaint is reviews that are too verbose. So it would be best if Navie can avoid giving a verbose answer when really there are no major concerns. 

Is Navie able to emit `N/A`, `OK`, or `No areas of substantial concern` for a particular section?

File: packages/navie/src/commands/review-command.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  ) {}
+
+  async *execute(req: CommandRequest): AsyncIterable<string> {
+    if (!req.codeSelection) {
+      // This should never happen.
+      yield 'An unknown error occurred. Please try again.';
+      return;
+    }
+
+    if (typeof req.codeSelection === 'string') {
```
Comment (kgilpin) [inline packages/navie/src/commands/review-command.ts:46]: This looks like something we could re-use elsewhere.

File: packages/navie/src/commands/review-command.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      yield 'Your `appmap` binary is out of date. Please update your client to the latest version.';
+      yield '';
+      yield 'Updates typically take place automatically. You can try restarting your code editor, or follow the manual instructions here:';
+      yield 'https://appmap.io/docs/reference/appmap-airgapped-install.html#download-the-appmap-application-binaries';
+      return;
+    }
+
+    const gitDiff = req.codeSelection.find(
+      (cs): cs is UserContext.CodeSnippetItem =>
+        cs.type === 'code-snippet' && cs.location === 'git diff'
```
Comment (kgilpin) [inline packages/navie/src/commands/review-command.ts:56]: Let's define `git diff` as a constant since it's being used in this structured way. 

Is it going to cause issues in the UI if `git diff` is interpreted as a location? Should we add a new kind of UserContext item type?

File: packages/navie/src/commands/review-command.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    );
+    if (!gitDiff) {
+      yield 'Unable to obtain the diff for the current branch. Please try again.';
```
Comment (kgilpin) [inline packages/navie/src/commands/review-command.ts:59]: It seems like we could do this at the CLI tier when obtaining the diff, and then just YOLO it here.

Comment (dustinbyrne) [reply packages/navie/src/commands/review-command.ts:59]: We can, and the client will generally throw an error if there's an issue obtaining the diff. That error will immediately be written out to the frontend as Navie's response, so no round trip will occur.

With that being said, this error is somewhat extraneous, but still possible due to the typing of `gitDiff` at this point (`ContextItem | undefined`). At least this case is covered should something change in the client or a bug is present capable of matching this condition.
